[Spec Decode] (1/2) Remove batch expansion

LiuXiaoxuanPKU opened the pull request:
This is a clean up PR for https://github.com/vllm-project/vllm/pull/5691 because that PR is stale. [RFC](https://docs.google.com/document/d/1I6H8bGYfwM8YappcP8HuoBhpS_pMlN0UIH2e_w9hv_M/edit)
Some constraints:
1. This PR current does not support cuda graph. It will be addressed in the follow up PR.
2. MQA scorer is only enabled for flash attention backend. 
3. It does not support ngram because MQA scorer does not handle mixed batches.

LiuXiaoxuanPKU committed:
draft

LiuXiaoxuanPKU committed:
w/o cuda graph support

LiuXiaoxuanPKU committed:
args and tests

LiuXiaoxuanPKU committed:
disable mqa for ngram and format

LiuXiaoxuanPKU committed:
clean up and tests

LiuXiaoxuanPKU committed:
revert example

LiuXiaoxuanPKU committed:
minor

LiuXiaoxuanPKU committed:
minor

comaniac reviewed (commented):
Overall LGTM.
Also cc @njhill  @sroy745 @cadedaniel

LiuXiaoxuanPKU committed:
fix tests -- chunked prefill and hiddens states in spec dec

LiuXiaoxuanPKU committed:
fix

LiuXiaoxuanPKU committed:
minor

LiuXiaoxuanPKU committed:
fix

LiuXiaoxuanPKU committed:
Merge branch 'main' into remove_batch_expansion

sroy745 commented:
> Overall LGTM. Also cc @njhill @sroy745 @cadedaniel

Thanks. Sorry could not take a look today. I will do it tomorrow.

sroy745 commented on vllm/attention/backends/flash_attn.py:
```diff
@@ -772,13 +788,17 @@ def forward(
             return prefill_output.view(num_prefill_tokens, hidden_size)
+
+        assert decode_meta is not None
+        assert decode_meta.decode_query_len == 1
```
We have this assert here because this case (chunked-prefill) won't happen for speculative decoding -- is my understanding correct? please consider adding a comment for this assert explaining why it would work for speculative decoding where decode_query_len might be > 1

LiuXiaoxuanPKU committed:
fix sampler for beam search

LiuXiaoxuanPKU committed:
revert num compute tokens

LiuXiaoxuanPKU committed:
disbale mqa scorer when draft model and target model have different model length

LiuXiaoxuanPKU committed:
diable mqa for cuda graph

LiuXiaoxuanPKU committed:
fix partial comments

LiuXiaoxuanPKU committed:
fix comments

LiuXiaoxuanPKU committed:
fix sampler and spec dec tests

LiuXiaoxuanPKU committed:
remove backend

LiuXiaoxuanPKU committed:
more test fix

LiuXiaoxuanPKU committed:
Merge branch 'main' into remove_batch_expansion

LiuXiaoxuanPKU committed:
fix num_compute_token

LiuXiaoxuanPKU commented:
Thanks for the review @comaniac @sroy745. I addressed the comments.The PR should be ready for review, please take another pass when you get time, thanks!

LiuXiaoxuanPKU committed:
clean up

comaniac commented on vllm/engine/llm_engine.py:
```diff
@@ -995,8 +995,11 @@ def _process_model_outputs(self,
+                    output_token_num = self.output_processor.process_outputs(
                         seq_group, output, is_async)
+                    if self.speculative_config:
+                        seq_group.update_num_computed_tokens(output_token_num -
+                                                             1)
```
I see. Could you add a comment?

comaniac commented on vllm/spec_decode/spec_decode_worker.py:
```diff
@@ -173,12 +176,37 @@ def create_worker(
+            "[Speculative Decoding] Configuring"
+            " SpecDecodeWorker with sampler=%s", type(spec_decode_sampler))
+
+        if scorer_worker.model_runner.attn_backend.get_name() != "flash-attn":
+            disable_mqa_scorer = True
+            logger.info("[Speculative Decoding] Disabling MQA scorer as the "
+                        "MQA is only available with flash attn backend.")
```
We should not show there messages if `disable_mqa_scorer` is true set by user.

comaniac reviewed (approved):
LGTM. How does this PR work with #8950?

LiuXiaoxuanPKU committed:
more fix for num_compute_token

LiuXiaoxuanPKU committed:
change log condition

LiuXiaoxuanPKU commented:
> LGTM. How does this PR work with #8950?

It's one addition based on #8950. #8950 removed the update in the decoding phase. I add it back, but add it in a different place as shown [here](https://github.com/LiuXiaoxuanPKU/vllm/blob/79ac29ce93aad4a59162683985dd6ef55961a6e7/vllm/engine/llm_engine.py#L1114) and [here](https://github.com/LiuXiaoxuanPKU/vllm/blob/79ac29ce93aad4a59162683985dd6ef55961a6e7/vllm/engine/llm_engine.py#L1264).

LiuXiaoxuanPKU committed:
add comments

LiuXiaoxuanPKU committed:
query len for multi-step, specify ci backend

LiuXiaoxuanPKU committed:
fix ci

LiuXiaoxuanPKU committed:
fix

LiuXiaoxuanPKU committed:
format

varun-sundar-rabindranath commented on vllm/attention/backends/blocksparse_attn.py:
```diff
@@ -186,6 +186,12 @@ class BlocksparseFlashAttentionMetadata(AttentionMetadata):
 
+    # Number of query tokens for each request in the batch.
+    # Currently, we require that all requests have the same number of query
+    # tokens during the decoding phase. When speculavie decoding is enabled,
```
nit: `speculavie` -> `speculative`

varun-sundar-rabindranath commented on vllm/engine/llm_engine.py:
```diff
@@ -1110,6 +1110,8 @@ def update_prefill_num_computed_tokens(
                                                    len(output),
                                                    is_first_step_output)
+            elif not is_async:
+                seq_group.update_num_computed_tokens(1)
```
There might be an issue here with multi-step + eager-mode - lets say there are 8 steps in a multi-step, when all the 8 outputs are submitted in a burst, process_model_outputs should account for all the 8 steps, but here we seem to account for only one ?

LiuXiaoxuanPKU committed:
context_len for multi-step and encoder decoder, fix decode_len
